src/sentry/event_manager.py

```diff
@@ -1891,6 +1891,55 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
     )
 
 
+def _is_stuck_counter_error(err: Exception, project: Project, short_id: int) -> bool:
+    """Decide if this is `UniqueViolation` error on the `Group` table's project and short id values."""
+
+    error_message = err.args[0]
+
+    for substring in [
+        f"Key (project_id, short_id)=({project.id}, {short_id}) already exists.",
+        'duplicate key value violates unique constraint "sentry_groupedmessage_project_id_short_id',
+    ]:
+        if substring in error_message:
+            return True
+
+    return False
+
+
+def _handle_stuck_project_counter(project: Project, current_short_id: int) -> int:
+    """
+    Sometimes, for reasons unknown, a project's `Counter` value falls behind its latest group id.
+    When that happens, that incorrect counter value leads us to try to create groups with ids which
+    are already taken.
+
+    This handles that case by updating the counter's value to the latest group id, and then returns
+    the new value.
+    """
+    new_short_id = current_short_id
+
+    # Ordinarily running max on this many rows would be prohibitively expensive, but a) this is
+    # a very rare case (< 20 ever that we know of), and b) project and short id are indexed
+    # together in order to enforce the unique constraint which got us here in the first place,
+    # so it's faster than it otherwise might be.
+    max_short_id_for_project = Group.objects.filter(project_id=project.id).aggregate(
+        Max("short_id")
+    )["short_id__max"]
+
+    # Add 1 because we're trying to mimic a value which would already have been incremented
+    correct_value = max_short_id_for_project + 1
+
+    if current_short_id != correct_value:
+        difference = correct_value - current_short_id
+        # `_get_next_short_id` corrects the `Counter` value before it returns the new short_id
+        new_short_id = _get_next_short_id(project, delta=difference)
+
+        logger.info(
+            "Fixed stuck counter value.", extra={"project": project.id, "difference": difference}
+        )
```

Comment:
I think we should write something to datadog here so we can easily alert on it and see how frequent it is (bump sampling to 100%)

Reply:
It'll have to be in addition rather than instead, yeah? Since we can't add the project id to the metric?

For now I'll go with that.

Reply:
Yeah, both is fine